File: src/linked-list/linked_list.py
"""A linked list in python."""


class Node(object):
    """."""

    def __init__(self, data, next):
        """Build node attributes."""
        self.data = data
        self.next = next


class LinkedList(object):
    """Create linked list object."""

    def __init__(self, iterable=None):
        """Head node is none on init."""
        self.head = None
        self._counter = 0
        if isinstance(iterable, (str, tuple, list)):
            for item in iterable:
                self.push(item)

    def push(self, data):
        """Push value to linked list."""
        new_node = Node(data, self.head)
        self.head = new_node
        self._counter += 1

    def pop(self):
        """Remove first item from list and return it."""
        if not self.head:
            raise IndexError("List is empty.")
        output = self.head.data
        self.head = self.head.next
        self._counter -= 1
        return output

    def size(self):
        """."""
        return self._counter

    def search(self, val):
        """."""
        curr = self.head
        while curr:
            if curr.data == val:
                return curr
            curr = curr.next

    def display(self):
        """."""
        output_string = "("
        node = self.head
        if node is None:
            return '()'
        while node:
            try:
                float(node.data)
                output_string += (str(node.data) + ", ")
            except ValueError:
                output_string += ("\'" + node.data + "\'") + ", "
            node = node.next
        output_string = output_string[:-2] + ')'
        return(output_string)

    def __len__(self):
        """Return size of linked list."""
        return self._counter

    def remove(self, node):
        """."""
        current_node = self.head
        previous_node = None
        while current_node:
            if current_node.data == node:
                if previous_node is None:
                    self.head = current_node.next
                    self._counter -= 1
                    return
                else:
                    previous_node.next = current_node.next
                    self._counter -= 1
                    return
            else:
                previous_node = current_node
                current_node = current_node.next
        raise IndexError('Node not in linked list.')

    def __str__(self):
        """Return what the display method returns."""
        return self.display()

File: js/linkedList.js
'use strict';

class Node {
    constructor(val, next=null) {
        this.val = val;
        this.next = next;
    }
};


class LinkedList {
    constructor(iterable=null) {
        this.head = null;
        this._length = 0;
        if(Array.isArray(iterable)) {
            for (let i = 0; i < iterable.length; i++) {
                this.push(iterable[i]);
            }
        }
    }

    push(val) {
        this.head = new Node(val, this.head);
        this._length++;
    }

    pop() {
        if (this._length === 0) {
            return null
        }

        if (this._length === 1) {
            let popped_val = this.head.val;
            this.head = null;
            this._length--;
            return popped_val
        }

        let node = this.head;
        while (node.next.next) {
            node = node.next;
        }
        let popped_val = node.next.val;
        node.next = null;
        this._length--;
        return popped_val;
    }

    size() {
        return this._length;
    }

    search(val) {
        let node = this.head;
        while (node) {
            if (node.val === val) {
                return node;
            }
            node = node.next;
        }
        return null;
    }

    remove(val) {
        if (this._length === 0) {
            throw new Error('List is empty.'); 
        }

        if (this.head.val === val) {
            this.head = this.head.next;
            this._length--;
            return
        }

        let node = this.head;
        while (node.next) {
            if (node.next.val === val) {
                node.next = node.next.next;
                this._length--;
                return;
            }
            node = node.next;
        }

        throw new Error('Value not in list');
    }

    display() {
        let output_string = '('
        let node = this.head;
        while (node) {
            if (typeof node.val === 'string' || node.val instanceof String)
                output_string += `\'${node.val}\', `
            else
                output_string += `${node.val}, `
            node = node.next;
        }
        output_string = output_string.slice(0, -2);
        return output_string += ')'
    }
};

module.exports = {LinkedList};